divfree/TPZCompElHCurlNoGrads.cpp

```diff
@@ -0,0 +1,369 @@
+  */
+  
+  //computes the index that will associate each scalar function to a constant vector field
+  constexpr auto nConnects = TSHAPE::NSides - TSHAPE::NCornerNodes;
+  TPZManVector<int,nConnects> connectOrders(nConnects,-1);
+  int unfiltnshape = 0;
+  for(auto i = 0; i < nConnects; i++){
+    const auto conorder = this->EffectiveSideOrder(i + TSHAPE::NCornerNodes);
+    connectOrders[i] = conorder;
+    unfiltnshape += TPZCompElHCurlFull<TSHAPE>::NConnectShapeF(i, conorder);
+  }
+
+
+  const auto nFaces = TSHAPE::Dimension < 2 ? 0 : TSHAPE::NumSides(2);
+  const auto nEdges = TSHAPE::NumSides(1);
+  constexpr auto nNodes = TSHAPE::NCornerNodes;
+
+  TPZManVector<int64_t,nNodes> nodes(nNodes, 0);
+
+  for (auto iNode = 0; iNode < nNodes; iNode++){
+    nodes[iNode] = this->Reference()->NodeIndex(iNode);
+  }
+
+  
+  TPZManVector<int64_t, TSHAPE::NSides - nNodes>
+    firstH1ShapeFunc(TSHAPE::NSides - nNodes,0);
+  
+  //calculates the first shape function associated with each side of dim > 0
+  TPZManVector<int,TSHAPE::NSides-nNodes> sidesH1Ord(TSHAPE::NSides - nNodes,-1);
+  this->CalcH1ShapeOrders(sidesH1Ord);
+  firstH1ShapeFunc[0] = nNodes;
+  
+  for (int iSide = nNodes + 1; iSide < TSHAPE::NSides; iSide++) {
+    const int iCon = iSide - nNodes;
+    firstH1ShapeFunc[iCon] =
+      firstH1ShapeFunc[iCon - 1] +
+      TSHAPE::NConnectShapeF(iSide - 1, sidesH1Ord[iCon-1]);
+
+  }
+
+  auto &indexVecShape = data.fVecShapeIndex;
+
+
+  
+  
+  indexVecShape.Resize(unfiltnshape);
+  TPZVec<unsigned int> shapeCountVec(TSHAPE::NSides - nNodes, 0);
+  this->StaticIndexShapeToVec(indexVecShape, connectOrders,
+                              firstH1ShapeFunc,sidesH1Ord, shapeCountVec, nodes);
+
+  //setting the type of shape functions as vector shape functions
+  data.fShapeType = TPZMaterialData::EVecShape;
+}
+
+template<class TSHAPE>
+template<class TVar>
+void TPZCompElHCurlNoGrads<TSHAPE>::ComputeRequiredDataT(
+  TPZMaterialDataT<TVar> &data, TPZVec<REAL> &qsi)
+{
+  
+  {
+    const bool needsSol = data.fNeedsSol;
+    data.fNeedsSol = false;
+    TPZIntelGen<TSHAPE>::ComputeRequiredData(data,qsi);//in this method, Shape will be called
+    data.fNeedsSol = needsSol;
+  }
+
+  this->ComputeDeformedDirections(data);    
+  /******************************************************************************************************************
+   * at this point, we already have the basis functions on the deformed element, since we have data.phi,
+   * data.fVecShapeIndex and data.fDeformedDirections. Now it is time to compute the curl, which will be stored in
+   * data.curlphi.
+   *******************************************************************************************************************/
+  TPZFMatrix<REAL> phiHCurl;
+  this->ComputeShape(data, phiHCurl);
+  
+
+  constexpr auto dim{TSHAPE::Dimension};
+  data.curlphi.Redim(2*dim - 3 > 0 ? 2*dim - 3 : 1, this->NShapeF());
+  this->ComputeCurl<dim>(data);
+  
```

Comment:
After computing the curl, the curl values should be copied to the vector values

Reply:
As of now, this is an HCurl element. In order to create the HDiv one, then yes, I think this would need to happen.